### PR TITLE
Repo changes - Added Add collection method

### PR DESCRIPTION
**Pull request by cthumburu**
Added Add collection method to the repository.

### CODE REVIEW

File: src/BoltOn.Data.EF/Repository.cs [hunk 1]
```diff
@@ -75,6 +74,22 @@ public virtual async Task DeleteAsync(TEntity entity, object options = null, Can
...
+			foreach (var entity in entities)
+			{
+				PublishEvents(entity);
+			}
+			await DbContext.SaveChangesAsync(cancellationToken);
+
+			return entities;
+		}
+
 		protected async Task SaveChangesAsync(TEntity entity, CancellationToken cancellationToken = default)
```
Comment (gokulm) [inline src/BoltOn.Data.EF/Repository.cs]: what if create another SaveChangesAsync method with IEnumerable < TEntity > and CancellationToken as params? and call it from the new AddAsync method?

the reason why i abstracted SaveChangesAsync and PublishEvents is for those who override the base repository can call SaveChangesAsync method and which will take care publishing events... ideally PublishEvents could be private. make all the SaveChanges and PublishEvents virtual.

File: tests/BoltOn.Tests/Data/CosmosDb/CosmosDbRepositoryTests.cs [hunk 2]
```diff
@@ -149,6 +150,27 @@ public async Task AddAsync_AddANewEntity_ReturnsAddedEntity()
 			Assert.Equal("meghan", result.FirstName);
 		}
 
+		[Fact]
+		public async Task AddAsync_AddANewEntities_ReturnsAddedEntities()
```
Comment (gokulm) [inline tests/BoltOn.Tests/Data/CosmosDb/CosmosDbRepositoryTests.cs]: remove A, AddAsync_AddNewEntities_ReturnsAddedEntities

File: tests/BoltOn.Tests/Data/EF/EFRepositoryTests.cs [hunk 3]
```diff
@@ -190,6 +190,34 @@ public async Task AddAsync_AddANewEntity_ReturnsAddedEntity()
 			Assert.Equal(result.FirstName, queryResult.FirstName);
 		}
 
+		[Fact, Trait("Category", "Integration")]
+		public async Task Add_AddANewEntities_ReturnsAddedEntities()
```
Comment (gokulm) [inline tests/BoltOn.Tests/Data/EF/EFRepositoryTests.cs]: above comment applies for this one too

File: src/BoltOn.Data.CosmosDb/Repository.cs [hunk 4]
```diff
@@ -118,7 +118,22 @@ public virtual async Task DeleteAsync(TEntity entity, object options = null, Can
 				await DocumentClient.DeleteDocumentAsync(GetDocumentUri(entityWithId.Id.ToString()));
 		}
 
-		protected Uri GetDocumentUri(string id)
+        public async Task<IEnumerable<TEntity>> AddAsync(IEnumerable<TEntity> entities, object options = null, CancellationToken cancellationToken = default)
+        {
+            foreach (var entity in entities)
+            {
+                PublishEvents(entity);
```
Comment (gokulm) [inline src/BoltOn.Data.CosmosDb/Repository.cs]: make the PublishEvents method virtual

File: src/BoltOn.Data.EF/Repository.cs (hunk 1 above)
Comment (cthumburu) [reply src/BoltOn.Data.EF/Repository.cs]: @gokulm If PublishEvents method is marked as virtual, then why making it private, it could remain protected.

Comment (gokulm) [reply src/BoltOn.Data.EF/Repository.cs]: @cthumburu do u see the need for making it protected virtual? if not, private should be good.

Comment (cthumburu) [reply src/BoltOn.Data.EF/Repository.cs]: @gokulm I thought PublishEvents was abstracted initially similar to SaveChangesAsync method for those who overrides the base repository and use the methods if needed. For SaveChangesAsync method, it make sense, I don't think anyone would use PublishEvents alone without persisting the changes. If that make sense, then I'll mark the method as private.

File: tests/BoltOn.Tests/Data/EF/EFRepositoryTests.cs [hunk 5]
```diff
@@ -190,6 +190,34 @@ public async Task AddAsync_AddANewEntity_ReturnsAddedEntity()
 			Assert.Equal(result.FirstName, queryResult.FirstName);
 		}
 
+		[Fact, Trait("Category", "Integration")]
+		public async Task AddAsync_AddANewEntities_ReturnsAddedEntities()
```
Comment (gokulm) [inline tests/BoltOn.Tests/Data/EF/EFRepositoryTests.cs]: @cthumburu  looks like you forgot to remove A from here AddNewEntitities

**Comment by gokulm** [comment]
@cthumburu i approved the PR. you can squash merge it. let the squash merge title be "feat: <add some title>" and the description be whatever you added to the PR description

**Pull request merged**